### PR TITLE
[Bug 1218563] Configure Docker nginx container to serve through https

### PR DESCRIPTION
**Pull request by safwanrahman**
This patch setups a Docker nginx container which makes the site serve through ssl with a self signed certificate. So the site can be accessed through `https://127.0.0.1`.
I have made a redirect rule in nginx so that all request coming from `http` will be redirected to `https`.

I think, some documentation change is needed. I will find out and fix them.

@jwhitlock r?

### CODE REVIEW

**Comment by safwanrahman** [comment]
I can not understand why the unit test consuming so much of the time after adding nginx. I will debug into it to understand where the issue is.

File: docker/images/nginx/Dockerfile [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+RUN mkdir /etc/nginx/ssl
+
+RUN openssl req \
+        -x509 \
+        -nodes \
+        -newkey rsa:2048 \
+        -keyout /etc/nginx/ssl/server.key \
+        -out /etc/nginx/ssl/server.crt \
+        -days 1000 \
+        -subj "/CN=127.0.0.1"
```
Comment (jgmize) [inline docker/images/nginx/Dockerfile]: I don't think we should include a private key in a  public docker image, even for 127.0.0.1. I believe a more correct approach, although less convenient, would be to document a version of this command (possibly wrapped in a "[make generate-cert](https://github.com/mdn/viewsourceconf/pull/98/files#diff-b67911656ef5d18c4ae36cb6741b7965R75)" or similar) to be run on the developer's laptop, and then [mount the cert and key](https://github.com/mdn/viewsourceconf/pull/98/files#diff-b67911656ef5d18c4ae36cb6741b7965R24) from the host at runtime.

Comment (safwanrahman) [inline docker/images/nginx/Dockerfile]: Cool. I will add it in next iteration.
Just debugging why the test taking a long time to run!

File: docker-compose.yml [hunk 2]
```diff
@@ -15,21 +15,20 @@ services:
...
       - ES_URLS=elasticsearch:9200
       - KUMASCRIPT_URL_TEMPLATE=http://kumascript:9080/docs/{path}
       - MEMCACHE_SERVERS=memcached:11211
-      - PROD_DETAILS_DIR=/app/product_details_json
-      - PROTOCOL=http://
-      - SESSION_COOKIE_SECURE=False
-      - SITE_URL=http://localhost:8000
+      - PROTOCOL=https://
+      - SESSION_COOKIE_SECURE=True
+      - SITE_URL=https://localhost:8000
```
Comment (jgmize) [inline docker-compose.yml]: The nginx container is listening on the standard port 443, so the `:8000` is incorrect here and may be causing timeouts during test runs.

File: docker/images/nginx/Dockerfile (hunk 1 above)
Comment (jgmize) [inline docker/images/nginx/Dockerfile]: Also, the generated cert is only valid for `127.0.0.1`; the `subj` should be `/CN=localhost`.

File: docker-compose.yml (hunk 2 above)
Comment (escattone) [inline docker-compose.yml]: Thanks jgmize! I made this change and, as you suspected, it resolved the timeouts that were occurring and consuming so much time when running the tests on this branch.

Comment (jgmize) [inline docker-compose.yml]: I'm glad that resolved the timeouts, but in hindsight I should have questioned the value of forcing tests through https by default anyway. Another option would be to use the `docker-compose.test.yml` file and continue to use http for these tests, and only use https for the integration tests.

**Comment by jwhitlock** [comment]
The environment updates have eliminated the timeouts that occurred with earlier versions of this PR.

We're now stuck with the "why" of the PR, which is less clear now.  Some possible whys:

* Persona has a requirement that the "audience" URL is an HTTPS URL (but Persona is going away)
* Browser security models prefer that SSL sites load SSL assets (but I believe HTTP sites can use HTTPS assets all day)
* Running in ``DEBUG=False`` mode turns on some checks that may fail if SSL is off
* It may be nice to check CSP headers in an HTTPS environment
* It would be nice to be able to run a production-like environment locally, including SSL serving of assets (but we're not sure what the production environment looks like yet)

In production, we'll probably:

1. Deploy web containers serving content over HTTP
2. Use AWS's [Elastic Load Balancer](https://aws.amazon.com/elasticloadbalancing/), with SSL termination at the the ELB, instead of nginx
3. Server static assets from the web containers or similar using [whitenoise](http://whitenoise.evans.io)
4. Use [CloudFront](https://aws.amazon.com/cloudfront/) to cache assets

Am I missing something? Should we put this PR on hold until we have a better idea of what the production environment is, and the benefits / costs of simulating it in the dev environment?

File: docker-compose.yml [hunk 3]
```diff
@@ -38,6 +37,7 @@ services:
   # Web is based on worker b/c you cannot clear the "ports" with docker-compose.
   web:
     <<: *worker
+    cpu_shares: 700
```
Comment (jwhitlock) [inline docker-compose.yml]: Is this still necessary or useful after the environment file changes?

**Comment by stephaniehobson** [comment]
There are some front-end stuff that behaves differently when you serve it over http vs https. I'd like to avoid getting mixed content errors in dev. It'll be hard to know what is and isn't a real security error.

Eventually we will want to add service workers to MDN, those can only be run on https.

**Comment by jwhitlock** [comment]
@stephaniehobson that sounds like a good reason to run with an SSL configuration for functional tests.  When do you need SSL in the dev environment?

**Comment by stephaniehobson** [comment]
@jwhitlock based on my work this sprint and our plans this quarter it can wait a few sprints. Also, I'm still developing in Vagrant (I know... I know...) so I can test https dependent things with it.

**Comment by jwhitlock** [comment]
It's been many months. We haven't needed this for testing, and service workers aren't on the roadmap.

@safwanrahman are you OK closing the PR? If not, please keep the branch up to date.

**Comment by escattone** [comment]
👍 I would vote to close it as well.

**Comment by safwanrahman** [comment]
@jwhitlock I am closing the PR for now. I will fix the conflict issue, and reopen this PR soon. Thanks a lot @escattone for the work. I am keeping the branch in my remote and upon fixing the issues, I will reopen it

**Pull request closed without merging**